python/lsst/ip/isr/ampOffset.py

```diff
@@ -18,88 +18,433 @@
+            below: TODO
+            -1: ...
+            ...: ...
+        ampSides : `numpy.ndarray`
+            An N x N matrix (N = the number of amplifiers) representing the amp
+            side information corresponding to the `ampAssociations`
+            matrix. The elements are integers defined as below:
+            -1: no side
+            0: side on the bottom
+            1: side on the right
+            2: side on the top
+            3: side on the left
+        """
+        xCenters = [amp.getBBox().getCenterX() for amp in amps]
+        yCenters = [amp.getBBox().getCenterY() for amp in amps]
+        xIndices = np.ceil(xCenters / np.min(xCenters) / 2).astype(int) - 1
+        yIndices = np.ceil(yCenters / np.min(yCenters) / 2).astype(int) - 1
+
+        nAmps = len(amps)
+        ampIds = np.zeros((len(set(yIndices)), len(set(xIndices))), dtype=int)
+
+        for ampId, xIndex, yIndex in zip(np.arange(nAmps), xIndices, yIndices):
+            ampIds[yIndex, xIndex] = ampId
+
+        ampAssociations = np.zeros((nAmps, nAmps), dtype=int)
+        ampSides = np.full_like(ampAssociations, -1)
+
+        for ampId in ampIds.ravel():
+            neighbors, sides = self.getNeighbors(ampIds, ampId)
+            ampAssociations[ampId, neighbors] = -1
+            ampSides[ampId, neighbors] = sides
+            ampAssociations[ampId, ampId] = -ampAssociations[ampId].sum()
+
+        if ampAssociations.sum() != 0:
+            raise RuntimeError("The `ampAssociations` array does not sum to zero.")
+
+        self.log.debug("amp associations:\n%s", ampAssociations)
+        self.log.debug("amp sides:\n%s", ampSides)
+
+        return ampAssociations, ampSides
+
+    def getNeighbors(self, ampIds, ampId):
+        """Get the neighbor amplifiers and their sides for a given
+        amplifier.
+
+        Parameters
+        ----------
+        ampIds : `numpy.ndarray`
+            Matrix with amp side association information.
+        ampId : `int`
+            The amplifier ID for which neighbor amplifiers and side IDs
+            are to be found.
+
+        Returns
+        -------
+        neighbors : `list` [`int`]
+            List of neighbor amplifier IDs.
+        sides : `list` [`int`]
+            List of side IDs, with each ID corresponding to its respective
+            neighbor amplifier.
+        """
+        m, n = ampIds.shape
+        r, c = np.ravel(np.where(ampIds == ampId))
+        neighbors, sides = [], []
+        sideLookup = {
+            0: (r + 1, c),
+            1: (r, c + 1),
+            2: (r - 1, c),
+            3: (r, c - 1),
+        }
+        for side, (row, column) in sideLookup.items():
+            if 0 <= row < m and 0 <= column < n:
+                neighbors.append(ampIds[row][column])
+                sides.append(side)
+        return neighbors, sides
+
+    def getInterfaceOffset(self, ampIdA, ampIdB, edgeA, edgeB):
```

Comment:
This is me being picky, but can this method be moved down after `getAmpEdges`?  This orders it after the method that calls it, and after the method that is called prior to it.

Reply:
Moved it.